### PR TITLE
[BEAM-1144] Spark runner fails to deserialize MicrobatchSource in cluster mode

### PR DESCRIPTION
**Pull request by aviemzur**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by aviemzur** [comment]
R: @amitsela @jbonofre

File: runners/spark/pom.xml [hunk 1]
```diff
@@ -134,14 +134,11 @@
...
     </dependency>
-    <!-- Kryo bugfix version needed due to a state re-use issue in Kryo version 2.21 used in Spark 1.x
-    See: https://issues.apache.org/jira/browse/SPARK-7708
-    See: https://github.com/EsotericSoftware/kryo/issues/312
-    -->
     <dependency>
       <groupId>com.esotericsoftware.kryo</groupId>
       <artifactId>kryo</artifactId>
-      <version>2.21.1</version>
+      <version>2.21</version>
```
Comment (amitsela) [inline runners/spark/pom.xml]: 👍

File: runners/spark/src/main/java/org/apache/beam/runners/spark/coders/BeamSparkRunnerRegistrator.java [hunk 2]
```diff
@@ -40,7 +40,7 @@
   @Override
   public void registerClasses(Kryo kryo) {
     for (Class<?> clazz : ClassesForJavaSerialization.getClasses()) {
-      kryo.register(clazz, new JavaSerializer());
+      kryo.register(clazz, new StatelessJavaSerializer());
```
Comment (amitsela) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/coders/BeamSparkRunnerRegistrator.java]: 👍

File: runners/spark/src/main/java/org/apache/beam/runners/spark/coders/ObjectInputStreamWithClassLoader.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    super(in);
+    this.classLoader = classLoader;
+  }
+
+  @Override
+  protected Class<?> resolveClass(ObjectStreamClass desc) {
+    try {
+      return Class.forName(desc.getName(), false, classLoader);
+    } catch (ClassNotFoundException e) {
+      throw new RuntimeException("Could not find class: " + desc.getName(), e);
```
Comment (amitsela) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/coders/ObjectInputStreamWithClassLoader.java]: What about `primeClasses` ?

File: runners/spark/src/main/java/org/apache/beam/runners/spark/coders/ObjectInputStreamWithClassLoader.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.ObjectInputStream;
+import java.io.ObjectStreamClass;
+
+
+/**
+ * ObjectInputStream with specific ClassLoader.
+ */
+class ObjectInputStreamWithClassLoader extends ObjectInputStream {
```
Comment (amitsela) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/coders/ObjectInputStreamWithClassLoader.java]: This is only used by `StatelessJavaSerializer`, why not make it an inner class ? 
`StatelessJavaSerializer` also provides explanations as to way this is needed at all.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/coders/StatelessJavaSerializer.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+import com.esotericsoftware.kryo.io.Output;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+
+
+/**
+ * Stateless Java Serializer.
+ *
+ * <p>
+ * Solves state re-use issue in Kryo version 2.21 used in Spark 1.x
```
Comment (amitsela) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/coders/StatelessJavaSerializer.java]: 👍

File: runners/spark/src/main/java/org/apache/beam/runners/spark/coders/StatelessJavaSerializer.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+ * See:
+ * https://issues.apache.org/jira/browse/SPARK-7708
+ * https://github.com/EsotericSoftware/kryo/issues/312
+ * </p>
+ *
+ * <p>
+ * Also, solves class loading issue in cluster caused by ${@link ObjectInputStream}
```
Comment (amitsela) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/coders/StatelessJavaSerializer.java]: 👍

File: runners/spark/src/test/java/org/apache/beam/runners/spark/coders/BeamSparkRunnerRegistratorTest.java [hunk 7]
```diff
@@ -49,9 +49,9 @@ public void testCodersAndSourcesRegistration() {
 
     for (Class<?> clazz : classesForJavaSerialization) {
       Assert.assertThat("Registered serializer for class " + clazz.getName()
-              + " was not an instance of " + JavaSerializer.class.getName(),
+              + " was not an instance of " + StatelessJavaSerializer.class.getName(),
```
Comment (amitsela) [inline runners/spark/src/test/java/org/apache/beam/runners/spark/coders/BeamSparkRunnerRegistratorTest.java]: Indents are 4 spaces usually.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/coders/StatelessJavaSerializer.java [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+ * ${@link ObjectInputStream} uses the last user-defined class loader in the stack which can be the
+ * wrong class loader.
+ * This is a known Java issue and a similar solution is often used.
+ * See:
+ * https://github.com/apache/spark/blob/v1.6.3/streaming/src/main/scala/org/apache/spark/streaming/Checkpoint.scala#L154
+ * https://issues.apache.org/jira/browse/GROOVY-1627
+ * https://github.com/spring-projects/spring-loaded/issues/107
+ * </p>
+ */
+class StatelessJavaSerializer extends Serializer {
```
Comment (amitsela) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/coders/StatelessJavaSerializer.java]: Maybe consider adding some comment on  `kryo.getGraphContext()` and why you're not using it..

File: runners/spark/src/main/java/org/apache/beam/runners/spark/coders/ObjectInputStreamWithClassLoader.java (hunk 4 above)
Comment (aviemzur) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/coders/ObjectInputStreamWithClassLoader.java]: 👍

File: runners/spark/src/main/java/org/apache/beam/runners/spark/coders/ObjectInputStreamWithClassLoader.java (hunk 3 above)
Comment (aviemzur) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/coders/ObjectInputStreamWithClassLoader.java]: `primClasses` is a private member of `ObjectInputStream`. In any case, primitives aren't serialized with `StatelessJavaSerializer` only the classes that were registered to use it (`Source` and `Coder` implementations)

File: runners/spark/src/test/java/org/apache/beam/runners/spark/coders/BeamSparkRunnerRegistratorTest.java (hunk 7 above)
Comment (aviemzur) [inline runners/spark/src/test/java/org/apache/beam/runners/spark/coders/BeamSparkRunnerRegistratorTest.java]: Will fixup

File: runners/spark/src/main/java/org/apache/beam/runners/spark/coders/ObjectInputStreamWithClassLoader.java (hunk 3 above)
Comment (amitsela) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/coders/ObjectInputStreamWithClassLoader.java]: 👍

**Comment by aviemzur** [comment]
@amitsela made the changes discussed in comments.

**Comment by amitsela** [comment]
LGTM from me, @jbonofre PTAL - specifically on `StatelessJavaSerializer` and it's `ObjectInputStreamWithClassLoader`.

**Comment by jbonofre** [comment]
CC: @jbonofre

Thanks guys ! Let me take a look on `ClassLoader`.

**Comment by aviemzur** [comment]
This change exists also on `Kryo` master: https://github.com/EsotericSoftware/kryo/commit/19a6b5edee7125fbaf54c64084a8d0e13509920b

**Comment by amitsela** [comment]
Thanks @aviemzur this is reassuring.

**Comment by jbonofre** [comment]
It sounds good to me, but I'm checking one more thing.

**Comment by amitsela** [comment]
Ping @jbonofre can we merge this ?

**Comment by amitsela** [comment]
Thanks!

**Pull request closed without merging**